NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java

```diff
@@ -0,0 +1,96 @@
+package dev.cobblesword.nachospigot.hitdetection;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import me.elier.nachospigot.config.NachoConfig;
+import org.apache.commons.lang3.tuple.Pair;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.util.Vector;
+
+import java.util.*;
+
+public class LagCompensator {
+
+    /*
+        Lag Compensator based on https://github.com/Islandscout/ServerSideHitDetection
+        Adjusts hit detection to be more fair for players with worse ping.
+     */
+
+    private final ListMultimap<UUID, Pair<Location, Long>> locationTimes;
+    private final int historySize;
+    private final int pingOffset;
+    private final int timeResolution;
+
+    public LagCompensator(int historySize, int pingOffset, int timeResolution) {
+        this.locationTimes = ArrayListMultimap.create();
+        this.historySize = historySize;
+        this.pingOffset = pingOffset;
+        this.timeResolution = timeResolution;
+    }
+
+    public Location getHistoryLocation(int rewindMillisecs, Player player) {
```

Comment:
This method is always called with the first arg being the players ping so this should be changed to remove the rewindMillisecs arg and make a variable for rewindMillisecs with the players ping. (`player.spigot().getPing()`) This also resolves Sculas’s comment about the cast optimization.